### ISSUE TITLE
Command "command" not found.

### ISSUE THREAD

**Issue by jayvdb** [coala/corobo#356]
We are regularly seeing `Command "foo" not found.` when the person correctly did `corobo foo ...`.
They used the wrong arguments, but the error returned is confusing.
It should tell them they gave the foo command the wrong arguments, and possible even force feed them the help in a DM to help them use the bot properly.

What is happening is that corobo is not doing partial matches on only the command, but the error is saying that it cant match the command.

```
> corobo lmgtfy
@co-robo: Command "lmgtfy" not found.
> corobo lmgtfy foo
Hey! You seem to ...
```

`lmgtfy` was the command in normal parlance.  `foo` is a parameter, or argument, etc.
The error message is wrong, and unhelpful.

**Comment by Makman2** [comment]
It's generelly advisable to use `arg_botcmd` and define a shell-like syntax :+1:

**Comment by Makman2** [comment]
Then you can usually post-process arguments but the command is still directed to the right function :+1:

**Comment by Vamshi99** [comment]
In my opinion, existing code is short and simple, using `arg_botcmd` makes the code complex.

**Comment by meetmangukiya** [comment]
I think this problems can be solved by not using `re_botcmd` everywhere. Only use it where it is extremely necessary. That way errbot has complete context and better metadata to report the errors appropriately.

**Comment by meetmangukiya** [comment]
I think this problems can be solved by not using `re_botcmd` everywhere. Only use it where it is extremely necessary. That way errbot has complete context and better metadata to report the errors appropriately.

**Comment by Makman2** [comment]
> In my opinion, existing code is short and simple, using arg_botcmd makes the code complex.

But when it doesn't match your regex again, we get this message another time ;)

**Comment by Makman2** [comment]
And `arg_botcmd` is actually very simple^^ It maybe adds 1 or 2 lines of overhead^^

**Comment by Vamshi99** [comment]
>But when it doesn't match your regex again, we get this message another time

Which message are you referring to?

**Comment by Makman2** [comment]
`Command "xyz" not found.`

**Comment by Vamshi99** [comment]
Yeah! that's what I almost fixed here, I am just getting some queue empty error here. Can you check my PR https://github.com/coala/corobo/pull/406

**Comment by Makman2** [comment]
> almost fixed 

That's why you should rather use `arg_botcmd` ;)

**Comment by Vamshi99** [comment]
Okay, the commit completely fixes the error, I just have some optional args error with `assign`and `pr stats` command`(it works perfectly with other commands). I request a review from you https://github.com/coala/corobo/pull/406

**Comment by Vamshi99** [comment]
See gitter chat at https://gitter.im/coala/corobo?at=5a2e4b2fc65707ba2b3995ef

Brief of discussion:
`arg_botcmd`: 
Disadvantages
* Help message will be missing
* Problem parsing the issue, pr links.
* Can't use aliases like (inv|invite|welcome)

Advantages:
* Stops every message being checked for matching regex.
* Would give better error checking.

**Comment by Makman2** [comment]
> Problem parsing the issue, pr links.

Why is that a problem? That step would happen in the command function^^

**Comment by Vamshi99** [comment]
@Makman2 If we use only `arg_botcmd` without regex matching, it would be a problem.

**Comment by Makman2** [comment]
```
@arg_botcmd(...)
def mycommand(..., arg1):
    import re
    m = re.match(myregex, arg1)
    if m is None:
        print('Error! arg1 is invalid!')
    else:
        # proceed
        pass
```
Doesn't appear to me as a limitation^^

> If we use only arg_botcmd without regex matching, it would be a problem.

You definitely have to be more precise^^

**Comment by Vamshi99** [comment]
Ok, I have agreed that way is good in gitter chat, what can we do with commands having aliases?

**Comment by Makman2** [comment]
We could copy function declarations and redirect to the main function in the worst case, but I don't really like it:
```python
# alias
@arg_botcmd(...)
def a(...):
    return assign(...)

@arg_botcmd(...)
def assign(...):
    # normal implementation
```

**Comment by Makman2** [comment]
I hope there's another way^^

**Comment by Vamshi99** [comment]
We will be having problem with help message as well, and I too don't like above method
